dts/arm/atmel/sam4l.dtsi

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2020 Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <arm/armv7-m.dtsi>
+#include <dt-bindings/i2c/i2c.h>
+#include <dt-bindings/gpio/gpio.h>
+
+#include "sam4l-pinctrl.dtsi"
+
+/ {
+	aliases {
+		watchdog0 = &wdog;
+	};
+
+	chosen {
+		zephyr,flash-controller = &flashcalw;
+		zephyr,entropy = &trng;
+	};
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu0: cpu@0 {
+			device_type = "cpu";
+			compatible = "arm,cortex-m4";
+			reg = <0>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			mpu: mpu@e000ed90 {
+				compatible = "arm,armv7m-mpu";
+				reg = <0xe000ed90 0x40>;
+				arm,num-mpu-regions = <8>;
+			};
+		};
+	};
+
+	soc {
+		sram0: memory@20000000 {
+			compatible = "mmio-sram";
+		};
+		/**
+		 * HRAM1 are 4k SRAM that can be used by PicoCache or
+		 * just extra memory available. The PicoCache will be
+		 * keep disable to ensures deterministic behaviour. That
+		 * way the extra memory can be exclusive for USB descriptors
+		 */
+		sram1: memory@21000000 {
+			compatible = "mmio-sram";
+			reg = <0x21000000 DT_SIZE_K(4)>;
+		};
+
+		flashcalw: flash-controller@400a0000 {
+			compatible = "atmel,sam-flash-controller";
+			label = "FLASH_CTRL";
+			reg = <0x400a0000 0x400>;
+			interrupts = <0 0>;
+			peripheral-id = <6>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			flash0: flash@0 {
+				compatible = "soc-nv-flash";
+				label = "FLASH_0";
+			};
+		};
+
+		wdog: watchdog@400f0c00 {
+			compatible = "atmel,sam-tw-watchdog";
+			reg = <0x400f0c00 0x400>;
+			interrupts = <44 0>;
+			peripheral-id = <163>;
+			label = "WATCHDOG_0";
+			psel = <16>;
+			status = "disabled";
+		};
+
+		spi0: spi@40008000 {
+			compatible = "atmel,sam-spi";
+			reg = <0x40008000 0x4000>;
+			interrupts = <54 0>;
+			peripheral-id = <65>;
+			label = "SPI_0";
+			status = "disabled";
+			#address-cells = <1>;
+			#size-cells = <0>;
+		};
+
+		usart0: usart@40024000 {
+			compatible = "atmel,sam-usart";
+			reg = <0x40024000 0x4000>;
+			interrupts = <65 1>;
+			peripheral-id = <72>;
```

Comment:
This is something that I can not change. SAM4L have a very different power management. On the below fragment  is possible see that rule will be different and the shift 6 doesn't apply here.
![image](https://user-images.githubusercontent.com/7027887/91915210-4b3dab80-ec90-11ea-8ee2-ce5ea0491e20.png)

Reply:
I still think the DTS should reflect the datasheet / header files. If we use different peripheral numbering scheme this will cause maintenance issues. The next person adding a sam4l peripheral to the DTS will use values from the header file. Also, someone struggling to bring up their module and seeing peripheral id not matching the header file will try to chase this trail even if the design was intentional.

It shouldn't take long for anyone to figure out that peripheral ids don't match data sheet on purpose but that's not a user friendly approach.

I would preserve peripheral ids as defined in the header files and modify `soc_pmc_peripheral_enable` function to decode bus_id and peripheral_bit values. Yes, the function code will be a bit more complex.

Reply:
I made the changes to allow the use of peripheral-id value, defined on the headers. Atmel defined start index (0) at PBAMASK increasing up to PBDMASK, then HSBMASK. For code consistency, I defined CPUMASK as index 5.  So, peripheral-id means at end the clock mask (bus clock) + bit index (peripheral).  Clock mask will be out of order on memory. To solve this, a map translate clock mask group to bus id.